core/src/main/java/bisq/core/payment/PaymentAccountManager.java

```diff
@@ -0,0 +1,128 @@
+        if (paymentAccount instanceof CryptoCurrencyAccount) {
+            CryptoCurrencyAccount cryptoCurrencyAccount = (CryptoCurrencyAccount) paymentAccount;
+            TradeCurrency tradeCurrency = cryptoCurrencyAccount.getSingleTradeCurrency();
+            if (tradeCurrency == null) {
+                throw new ValidationException("CryptoCurrency account must have exactly one trade currency");
+            }
+            altCoinAddressValidator.setCurrencyCode(tradeCurrency.getCode());
+            InputValidator.ValidationResult validationResult = altCoinAddressValidator.validate(cryptoCurrencyAccount.getAddress());
+            if (!validationResult.isValid) {
+                throw new ValidationException(validationResult.errorMessage);
+            }
+        }
+//        TODO we should validate payment account here as well
+        user.addPaymentAccount(paymentAccount);
+        TradeCurrency singleTradeCurrency = paymentAccount.getSingleTradeCurrency();
+        List<TradeCurrency> tradeCurrencies = paymentAccount.getTradeCurrencies();
+        if (singleTradeCurrency != null) {
+            if (singleTradeCurrency instanceof FiatCurrency)
+                preferences.addFiatCurrency((FiatCurrency) singleTradeCurrency);
+            else
+                preferences.addCryptoCurrency((CryptoCurrency) singleTradeCurrency);
+        } else if (!tradeCurrencies.isEmpty()) {
+            tradeCurrencies.forEach(tradeCurrency -> {
+                if (tradeCurrency instanceof FiatCurrency)
+                    preferences.addFiatCurrency((FiatCurrency) tradeCurrency);
+                else
+                    preferences.addCryptoCurrency((CryptoCurrency) tradeCurrency);
+            });
+        }
+        if (!(paymentAccount instanceof CryptoCurrencyAccount)) {
+            if (singleTradeCurrency == null && !tradeCurrencies.isEmpty()) {
+                if (tradeCurrencies.contains(CurrencyUtil.getDefaultTradeCurrency()))
+                    paymentAccount.setSelectedTradeCurrency(CurrencyUtil.getDefaultTradeCurrency());
+                else
+                    paymentAccount.setSelectedTradeCurrency(tradeCurrencies.get(0));
+
+            }
+            accountAgeWitnessService.publishMyAccountAgeWitness(paymentAccount.getPaymentAccountPayload());
+        }
+        return paymentAccount;
+    }
+
+    public void removePaymentAccount(String id) {
+        PaymentAccount paymentAccount = user.getPaymentAccount(id);
+        if (paymentAccount == null) {
+            throw new NotFoundException(format("Payment account %s not found", id));
+        }
+        boolean isPaymentAccountUsed = openOfferManager.getObservableList().stream()
+                .anyMatch(openOffer -> id.equals(openOffer.getOffer().getMakerPaymentAccountId()));
+        if (isPaymentAccountUsed) {
+            throw new PaymentAccountInUseException(format("Payment account %s is used for open offer", id));
```

Comment:
I prefer to use exceptions only in exceptional cases not in valid use case (user tries to remove account but app does not allow it as its used in trades or offers). Maybe best to refactor the check out to 'canRemove()' method and only call the method if it is allowed to remove.

Reply:
No! The method must not make any assumptions that it is being called in happy path scenario.
Exceptions are exactly for this type of situations where something else needs to be signaled then the happy path result.

Reply:
I disagree. I think the method removePaymentAccount should do what its name suggests and not throw an unchecked exception if the valid case happens that the user tries to remove an account but has an offer open using it. 
Bisq is using exceptions for exceptional cases not for flow control. I think we should stick with that patter and not introduce tons of exceptions for all kind of checks.

Reply:
I see this exception is handled [here](https://github.com/blabno/exchange/blob/a98cf1b9e84ceca658692547a92b0ec45653460b/desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java#L100-L107). Maybe `removePaymentAccount` method can return boolean directly ?

Reply:
Not using exceptions for flow control? What should they be used for then?

Reply:
> Not using exceptions for flow control? What should they be used for then?

This is considered bad practice:
https://duckduckgo.com/?q=java+exceptions+for+flow+control&t=ffab&ia=web

Reply:
> What should they be used for then?

For cases which happens exceptionally and sometimes lack of a clear way to handle it, so you delegate it to higher level classes how to deal with it. Writing to disk but get a IOError for instance. This use case is a valid circumstance that a user want to remove an account but does not know that there is a policy that he cannot remove it if used in offers or trades. This is not exceptional. I guess you want to provide information for the caller about the reason why the remove was not successful in the exception. You could also create a resultObject which contains diff. reuslt types and meta data if needed. Or just let the client call some validation methods like `isAccountUsedInOpenOffers` which is the usual way how Bisq handles such cases.

Reply:
So instead of of throwing Exception we should return some Object that either contains happy path result, or some additional data in case of failure?
Exceptions are much cleaner.

Reply:
I give up.... 
NACK to the PR

Reply:
Yes, give up, cause you don't have better arguments. When there are many reasons why a method might fail then the typed exceptions are the way to inform other component about the reason.
Strings are good for humans, but it's not only the humans that interact with this business logic component.

Reply:
Why strings? You can create some kind of OperationResult class, like [ValidationResult](https://github.com/bisq-network/bisq/blob/049655b6857449aaca789c9cee5358daf0a54419/core/src/main/java/bisq/core/util/validation/InputValidator.java#L37-L64).

You'll be able to map it to appropriate response in API.

Reply:
@lusarz do you honestly use OperationResult for every non-void method that you write?

Reply:
No, and I don't have strong opinion what is better - Exceptions or ResultObject. I usually use Exceptions, but now I see a lot of opinions in net that recommends ResultObject. I think this approach might be interesting. Client code with a lot of `catch` blocks doesn't looks too nice. 

Just give it a try, maybe you'll find some pros :)

Also consistency argument is relevant here. Validation in bisq [(for example here)](https://github.com/bisq-network/bisq/blob/049655b6857449aaca789c9cee5358daf0a54419/core/src/main/java/bisq/core/util/validation/InputValidator.java#L26-L28) uses ResultObject instead of throwing typed exceptions.

Reply:
Validation is completely different case. When you call `validate` method the main thing you want to know is the result of the validation.
Look at classic example of ArrayList: http://www.docjar.com/html/api/java/util/ArrayList.java.html#441

Does remove method return RemoveMethodResult? No it returns the removed object, which is the main intended thing. In order to tell user that object cannot be removed, because index is wrong it uses IndexOutOfBoundsException.

Reply:
Look at ArrayList [30 lines lower](http://www.docjar.com/html/api/java/util/ArrayList.java.html#472), it has `remove(Object o)` method. Does it throw `NotFoundException` ? No, it returns false.